pinot-core/src/main/java/org/apache/pinot/core/operator/transform/transformer/datetime/BaseDateTimeTransformer.java

```diff
@@ -58,41 +74,145 @@ public BaseDateTimeTransformer(@Nonnull DateTimeFormatSpec inputFormat, @Nonnull
+            dateTime.minuteOfHour().roundFloor();
+            return print(dateTime);
+          };
+          _millisDateTimeTruncate = (dateTime) -> {
+            dateTime.setMinuteOfHour((dateTime.getMinuteOfHour() / size) * size);
+            dateTime.minuteOfHour().roundFloor();
+            return dateTime.getMillis();
+          };
+        } else {
+          _sdfDateTimeTruncate = (dateTime) -> {
+            dateTime.minuteOfHour().roundFloor();
+            return print(dateTime);
+          };
+          _millisDateTimeTruncate = (dateTime) -> {
+            dateTime.minuteOfHour().roundFloor();
+            return dateTime.getMillis();
+          };
+        }
         break;
       case HOURS:
-        _dateTimeTruncate = (dateTime) -> _outputDateTimeFormatter
-            .print(dateTime.withHourOfDay((dateTime.getHourOfDay() / sz) * sz).hourOfDay().roundFloorCopy());
+        if (size != 1) {
+          _sdfDateTimeTruncate = (dateTime) -> {
+            dateTime.setHourOfDay((dateTime.getHourOfDay() / size) * size);
+            dateTime.hourOfDay().roundFloor();
+            return print(dateTime);
+          };
+          _millisDateTimeTruncate = (dateTime) -> {
+            dateTime.setHourOfDay((dateTime.getHourOfDay() / size) * size);
+            dateTime.hourOfDay().roundFloor();
+            return dateTime.getMillis();
+          };
+        } else {
+          _sdfDateTimeTruncate = (dateTime) -> {
+            dateTime.hourOfDay().roundFloor();
+            return print(dateTime);
+          };
+          _millisDateTimeTruncate = (dateTime) -> {
+            dateTime.hourOfDay().roundFloor();
+            return dateTime.getMillis();
+          };
+        }
         break;
       case DAYS:
-        _dateTimeTruncate = (dateTime) -> _outputDateTimeFormatter
-            .print(dateTime.withDayOfMonth(((dateTime.getDayOfMonth() - 1) / sz) * sz + 1).dayOfMonth()
-                .roundFloorCopy());
+        if (size != 1) {
+          _sdfDateTimeTruncate = (dateTime) -> {
+            dateTime.setDayOfMonth(((dateTime.getDayOfMonth() - 1) / size) * size + 1);
+            dateTime.dayOfMonth().roundFloor();
+            return print(dateTime);
+          };
+          _millisDateTimeTruncate = (dateTime) -> {
+            dateTime.setDayOfMonth(((dateTime.getDayOfMonth() - 1) / size) * size + 1);
+            dateTime.dayOfMonth().roundFloor();
+            return dateTime.getMillis();
+          };
+        } else {
+          _sdfDateTimeTruncate = (dateTime) -> {
+            dateTime.dayOfMonth().roundFloor();
+            return print(dateTime);
+          };
+          _millisDateTimeTruncate = (dateTime) -> {
+            dateTime.dayOfMonth().roundFloor();
+            return dateTime.getMillis();
+          };
+        }
         break;
       default:
-        _dateTimeTruncate = _outputDateTimeFormatter::print;
+        _sdfDateTimeTruncate = _outputDateTimeFormatter::print;
```

Comment:
Should we throw exception here?

Reply:
The other possible values are MICROSECONDS and NANOSECONDS so truncating to them is a no-op  as long as size is not > 1000.

Reply:
I see. Let's add some comments explaining this

Reply:
Done